### PR TITLE
[Lang] Support TensorType for irpass::alg_simp()

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: #

### <samp>🤖 Generated by Copilot at 411b19e</samp>

Add a template function `get_const_stmt_with_value` to `taichi/ir/statements.h` that creates IR statements for constants of different types. This is part of improving matrix support in the IR.

### <samp>🤖 Generated by Copilot at 411b19e</samp>

*  Add a new template function `get_const_stmt_with_value` to create constant statements of different types ([link](https://github.com/taichi-dev/taichi/pull/8225/files?diff=unified&w=0#diff-917d9436dcaafa0f1e41ae9bad90273a303f036f00da94e417788a7fa1dc5260R2021-R2050))

### CODE REVIEW

File: taichi/transforms/alg_simp.cpp [hunk 1]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
+                  is_integral(stmt->ret_type.get_element_type())) &&
                  irpass::analysis::same_value(stmt->lhs, stmt->rhs)) {
         // fast_math or integral operands: a -^ a -> 0
         replace_with_zero(stmt);
       }
-    } else if (rhs && stmt->op_type == BinaryOpType::pow) {
-      float64 exponent = rhs->val.val_cast_to_float64();
-      if (exponent == 1) {
+    } else if (stmt->op_type == BinaryOpType::pow) {
+      std::cout << 111111 << std::endl;
```
Comment (ailzhang) [inline taichi/transforms/alg_simp.cpp:495]: nit: need remove

**Review by ailzhang — APPROVED** [review]

File: tests/python/test_ad_basics.py [hunk 2]
```diff
@@ -255,15 +255,15 @@ def test_atan2_f64(tifunc, npfunc):
 @pytest.mark.parametrize(
     "tifunc,npfunc",
     [
-        (lambda x: 0.4**x, lambda x: np.power(0.4, x)),
+        # (lambda x: 0.4**x, lambda x: np.power(0.4, x)),
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply tests/python/test_ad_basics.py:258]: thx!

File: taichi/transforms/alg_simp.cpp [hunk 3]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
+      if (exponent_one_optimize(stmt)) {
         // a ** 1 -> a
-        stmt->replace_usages_with(stmt->lhs);
-        modifier.erase(stmt);
-      } else if (exponent == 0) {
+        std::cout << 2222 << std::endl;
```
Comment (dream189free) [inline taichi/transforms/alg_simp.cpp:497]: ditto~

File: taichi/transforms/alg_simp.cpp [hunk 4]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
+      } else if (exponent_zero_optimize(stmt)) {
         // a ** 0 -> 1
-        replace_with_one(stmt);
-      } else if (exponent == 0.5) {
+        std::cout << 3333 << std::endl;
```
Comment (dream189free) [inline taichi/transforms/alg_simp.cpp:500]: ditto~

File: taichi/transforms/alg_simp.cpp [hunk 5]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
-        auto a = stmt->lhs;
-        cast_to_result_type(a, stmt);
-        auto result = Stmt::make<UnaryOpStmt>(UnaryOpType::sqrt, a);
-        result->ret_type = a->ret_type;
-        stmt->replace_usages_with(result.get());
-        modifier.insert_before(stmt, std::move(result));
-        modifier.erase(stmt);
-      } else if (exponent == std::round(exponent) && exponent > 0 &&
-                 exponent <= max_weaken_exponent) {
+        std::cout << 4444 << std::endl;
```
Comment (dream189free) [inline taichi/transforms/alg_simp.cpp:503]: ditto~

File: taichi/transforms/alg_simp.cpp [hunk 6]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
-              BinaryOpType::mul, a_power_of_2, a_power_of_2);
-          new_a_power->ret_type = a->ret_type;
-          a_power_of_2 = new_a_power.get();
-          modifier.insert_before(stmt, std::move(new_a_power));
-        }
-        stmt->replace_usages_with(result);
-        modifier.erase(stmt);
-      } else if (exponent == std::round(exponent) && exponent < 0 &&
-                 exponent >= -max_weaken_exponent) {
+        std::cout << 5555 << std::endl;
```
Comment (dream189free) [inline taichi/transforms/alg_simp.cpp:506]: ditto~

File: taichi/transforms/alg_simp.cpp [hunk 7]
```diff
@@ -244,82 +485,29 @@ class AlgSimp : public BasicStmtVisitor {
...
-                                               new_exponent.get());
-        a_to_n->ret_type = stmt->ret_type;
-        auto result =
-            Stmt::make<BinaryOpStmt>(BinaryOpType::div, one_raw, a_to_n.get());
-        stmt->replace_usages_with(result.get());
-        modifier.insert_before(stmt, std::move(new_exponent));
-        modifier.insert_before(stmt, std::move(a_to_n));
-        modifier.insert_before(stmt, std::move(result));
-        modifier.erase(stmt);
+        std::cout << 6666 << std::endl;
```
Comment (dream189free) [inline taichi/transforms/alg_simp.cpp:509]: ditto~

File: taichi/transforms/alg_simp.cpp (hunk 3 above)
Comment (jim19930609) [reply taichi/transforms/alg_simp.cpp:497]: thx~

**Review by dream189free — APPROVED** [review]
LGTM!

**Pull request merged**